proposals/default_optional_arguments/revised_proposal.txt

```diff
@@ -0,0 +1,356 @@
+  * Use a separate variable inside the procedure to assign the value
+    because the optional dummy argument that is not present must not
+    be referenced in expressions other than as actual argument to the
+    intrinsic function present().
+
+This example function illustrates the problem:
+
+    real function quadratic(x, a, b, c)  
+      ! returns a + b * x + c * x**2 if c is present 
+      ! and a + b * x otherwise
+      real, intent(in) :: x, a, b
+      real, intent(in), optional :: c
+      real :: c_tmp ! use another var. to reference the missing arg
+      c_tmp = 0 ! default value if c is not present
+      if (present(c)) c_tmp = c
+      quadratic = a + b * x + c_tmp * x**2
+    end function quadratic
+
+For any dummy argument with the optional attribute, the programmer
+must use the intrinsic function present() to check for the presence of
+the argument. Furthermore, for INTENT(IN) or VALUE or no intent
+arguments, if the optional dummy argument is meant to be used in
+multiple places in the procedure, the programmer is likely to use the
+above pattern, where a "temporary" variable is declared and used in
+place of the dummy argument, disconnecting the implementation from the
+user interface. Furthermore, this requires at least 3 lines of code
+(declaration of c_tmp, initialization of c_tmp, and testing for the
+presence of c) to handle the scenario of a missing optional argument.
+
+The situation for other intents is different. For INTENT(OUT) one
+might like to have a default value, but there is no need for a
+temporary variable to hold it. The example subroutine shows a typical
+usage:
+
+    subroutine do_stuff( ..., status )
+	...
+	integer, intent(out), optional :: status
+	...
+	if (present(status)) status = 0 ! the desired default
+	...
+	if (bad_stuff_happened) then
+	    if (present(status)) then
+	        status = 1 ! Flag that bad stuff happened
+		return
+	    else
+	        error stop 'Bad Stuff Happened in DO_STUFF'
+            end if
+	end if
+	...
+    end subroutine do_stuff
+
+This differs from the INTENT(IN), VALUE, and no intent case: first,
+because one assigns the default value with INTENT(OUT) when the actual
+argument is present, not when it is absent; and second, because the
+assignment of a default involves only one statement rather than the
+three statements for the helper variable of INTENT(IN), VALUE, and no
+intent. For INTENT(INOUT), one typically makes the argument optional
+to avoid the computational expense of calculating the out result, but
+having a default value makes it awkward to determine the presence of
+an optional argument, so the primary purpose of having it be optional
+is defeated.
+
+This proposal addresses the issue for INTENT(IN), VALUE, and no intent
+arguments that checking for the presence of the optional dummy
+argument and using a helper variable is cumbersome and error-prone. We
+will not address default values for INTENT(OUT) OR INTENT(INOUT)
```

Comment:
I am not sure to understand why `intent(inout)` is not considered.If no argument is passed, then `intent(inout)` would behave as `intent(in)`?